Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+                // If next() hasn't moved then we have read through all the
+                // directories, and can exit.
+                return KSuccess;
+            }
+
+            continue;
+        } else if (result == RecursionDecision::Break) {
+            return KSuccess;
+        } else {
+            while (!iterator.done()) {
+                if (!iterator.skip()) {
+                    if (!iterator.go_up()) {
+                        return KSuccess;
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    return KSuccess;
+}
+
+KResultOr<NonnullRefPtr<ISO9660FS::DirectoryEntry>> ISO9660FS::directory_entry_for_record(Badge<ISO9660DirectoryIterator>, ISO::DirectoryRecordHeader const* record)
+{
+    u32 extent_location = LittleEndian { record->extent_location.little };
+    u32 data_length = LittleEndian { record->data_length.little };
+
+    auto key = calculate_directory_entry_cache_key(record);
+    auto it = m_directory_entry_cache.find(key);
+    if (it != m_directory_entry_cache.end()) {
+        dbgln_if(ISO9660_DEBUG, "Cache hit for dirent @ {}", extent_location);
+        return it->value;
+    }
+    dbgln_if(ISO9660_DEBUG, "Cache miss for dirent @ {} :^(", extent_location);
+
+    if (m_directory_entry_cache.size() == max_cached_directory_entries) {
+        // FIXME: A smarter algorithm would probably be nicer.
+        m_directory_entry_cache.remove(m_directory_entry_cache.begin());
+    }
+
+    if (!(data_length % logical_block_size() == 0)) {
+        dbgln_if(ISO9660_DEBUG, "Found a directory with non-logical block size aligned data length!");
+        return EIO;
+    }
+
+    auto blocks = KBuffer::try_create_with_size(data_length, Region::Access::Read | Region::Access::Write, "ISO9660FS: Directory traversal buffer");
+    if (!blocks) {
+        return ENOMEM;
+    }
+
+    auto blocks_buffer = UserOrKernelBuffer::for_kernel_buffer(blocks->data());
+    auto did_read = raw_read_blocks(BlockBasedFileSystem::BlockIndex { extent_location }, data_length / logical_block_size(), blocks_buffer);
+    if (!did_read) {
+        return EIO;
+    }
+
+    auto maybe_entry = DirectoryEntry::create(extent_location, data_length, move(blocks));
+    if (maybe_entry.is_error()) {
+        return maybe_entry.error();
+    }
+    m_directory_entry_cache.set(key, maybe_entry.value());
+
+    dbgln_if(ISO9660_DEBUG, "Cached dirent @ {}", extent_location);
+    return maybe_entry.release_value();
+}
+
+u32 ISO9660FS::calculate_directory_entry_cache_key(ISO::DirectoryRecordHeader const* record)
+{
+    return LittleEndian { record->extent_location.little };
+}
+
+KResultOr<size_t> ISO9660Inode::read_bytes(off_t offset, size_t size, UserOrKernelBuffer& buffer, FileDescription*) const
+{
+    MutexLocker inode_locker(m_inode_lock);
+
+    auto& file_system = const_cast<ISO9660FS&>(static_cast<ISO9660FS const&>(fs()));
```

Comment:
I assume this is because of the cache?
pls make the cache `mutable` and keep this `const`, if that's the issue.

Reply:
No, this is because `raw_read` from `BlockBasedFileSystem` is unfortunately a non-`const` operation.

Reply:
Hm.
Is _that_ because it wants to populate a cache?

Reply:
Not quite, it's because that reads from the `file_description()`.

Reply:
Shame.
Alright